File: src/weaponsprocurement/gui/WimGuiText.java
package weaponsprocurement.gui;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

final class WimGuiText {
    private static final Set<String> WEAK_LINE_ENDINGS = new HashSet<String>(Arrays.asList(
            "a", "an", "and", "as", "at", "by", "for", "from", "in", "of", "or", "the", "to", "with"));

    private WimGuiText() {
    }

    static String fit(String text, int maxVisibleChars) {
        return truncateLineToFit(text == null ? "" : text, maxVisibleChars, 10);
    }

    static int estimatedChars(float width) {
        return estimatedChars(width, 0f, WimGuiStyle.TEXT_APPROX_CHAR_WIDTH);
    }

    static int estimatedChars(float width, float horizontalPadding, float approxCharWidth) {
        float effectiveWidth = Math.max(8f, width - Math.max(0f, horizontalPadding));
        return Math.max(1, (int) (effectiveWidth / Math.max(1f, approxCharWidth)));
    }

    static List<String> wrap(String text, int maxCharsPerLine, int maxLines) {
        int safeChars = Math.max(1, maxCharsPerLine);
        int safeLines = Math.max(1, maxLines);
        List<String> lines = new ArrayList<String>();
        String[] rawLines = (text == null ? "" : text).split("\\r?\\n", -1);
        for (int i = 0; i < rawLines.length; i++) {
            lines.addAll(wrapTextLineWordAware(rawLines[i], safeChars));
        }
        if (lines.isEmpty()) {
            lines.add("");
        }
        if (lines.size() <= safeLines) {
            return lines;
        }
        List<String> visible = new ArrayList<String>(lines.subList(0, safeLines));
        StringBuilder overflow = new StringBuilder();
        for (int i = safeLines - 1; i < lines.size(); i++) {
            if (overflow.length() > 0) {
                overflow.append(' ');
            }
            overflow.append(lines.get(i));
        }
        visible.set(visible.size() - 1, fit(overflow.toString(), safeChars));
        return visible;
    }

    static WimGuiTextLayout fitLayout(String text,
                                      float availableWidth,
                                      float availableHeight,
                                      float minRowHeight,
                                      float horizontalPadding,
                                      float verticalPadding,
                                      float approxCharWidth,
                                      float lineHeight,
                                      int maxLines,
                                      boolean canGrowHeight,
                                      float maxHeight) {
        int safeMaxLines = Math.max(1, maxLines);
        float safeLineHeight = Math.max(1f, lineHeight);
        int maxCharsPerLine = estimatedChars(availableWidth, horizontalPadding, approxCharWidth);
        int heightLimitedLines = safeMaxLines;
        if (!canGrowHeight && !Float.isInfinite(availableHeight)) {
            heightLimitedLines = Math.max(1, (int) ((availableHeight - verticalPadding) / safeLineHeight));
        }
        int cappedMaxLines = Math.max(1, Math.min(safeMaxLines, heightLimitedLines));
        List<String> wrappedLines = wrapAll(text, maxCharsPerLine);
        boolean truncated = wrappedLines.size() > cappedMaxLines;
        List<String> cappedLines = new ArrayList<String>();
        if (truncated) {
            cappedLines.addAll(wrappedLines.subList(0, cappedMaxLines));
            StringBuilder overflow = new StringBuilder();
            for (int i = cappedMaxLines - 1; i < wrappedLines.size(); i++) {
                if (overflow.length() > 0) {
                    overflow.append(' ');
                }
                overflow.append(wrappedLines.get(i));
            }
            cappedLines.set(cappedLines.size() - 1, fit(overflow.toString(), maxCharsPerLine));
        } else {
            cappedLines.addAll(wrappedLines);
        }
        if (cappedLines.isEmpty()) {
            cappedLines.add("");
        }

        int lineCount = Math.max(1, cappedLines.size());
        float desiredHeight = Math.max(minRowHeight, verticalPadding + lineCount * safeLineHeight);
        float rowHeight = canGrowHeight
                ? Math.min(desiredHeight, maxHeight)
                : Math.min(desiredHeight, availableHeight);
        int widest = 0;
        for (int i = 0; i < cappedLines.size(); i++) {
            widest = Math.max(widest, cappedLines.get(i).length());
        }
        return new WimGuiTextLayout(
                joinLines(cappedLines),
                lineCount,
                rowHeight,
                Math.min(Math.max(8f, availableWidth - horizontalPadding), widest * approxCharWidth),
                rowHeight,
                truncated);
    }

    static WimGuiTextLayout fitLayout(String text, float width, float minRowHeight, int maxLines) {
        return fitLayout(
                text,
                width,
                Float.POSITIVE_INFINITY,
                minRowHeight,
                WimGuiStyle.TEXT_LEFT_PAD,
                WimGuiStyle.TEXT_VERTICAL_PADDING,
                WimGuiStyle.TEXT_APPROX_CHAR_WIDTH,
                WimGuiStyle.TEXT_LINE_HEIGHT,
                maxLines,
                true,
                Float.POSITIVE_INFINITY);
    }

    private static List<String> wrapAll(String text, int maxCharsPerLine) {
        int safeChars = Math.max(1, maxCharsPerLine);
        List<String> lines = new ArrayList<String>();
        String[] rawLines = (text == null ? "" : text).split("\\r?\\n", -1);
        for (int i = 0; i < rawLines.length; i++) {
            lines.addAll(wrapTextLineWordAware(rawLines[i], safeChars));
        }
        return lines.isEmpty() ? Collections.singletonList("") : lines;
    }

    private static List<String> wrapTextLineWordAware(String line, int maxCharsPerLine) {
        if (line == null || line.trim().isEmpty()) {
            return Collections.singletonList("");
        }
        if (!line.contains(" ")) {
            return wrapLongToken(line.trim(), maxCharsPerLine);
        }

        List<String> wrapped = new ArrayList<String>();
        String current = "";
        String[] words = line.trim().split("\\s+");
        for (int i = 0; i < words.length; i++) {
            List<String> segments = words[i].length() > maxCharsPerLine
                    ? wrapLongToken(words[i], maxCharsPerLine)
                    : Collections.singletonList(words[i]);
            for (int segmentIndex = 0; segmentIndex < segments.size(); segmentIndex++) {
                String segment = segments.get(segmentIndex);
                if (current.isEmpty()) {
                    current = segment;
                } else if (current.length() + 1 + segment.length() <= maxCharsPerLine) {
                    current = current + " " + segment;
                } else {
                    wrapped.add(current);
                    current = segment;
                }
                if (segmentIndex < segments.size() - 1) {
                    wrapped.add(current);
                    current = "";
                }
            }
        }
        if (!current.isEmpty()) {
            wrapped.add(current);
        }
        return wrapped.isEmpty() ? Collections.singletonList("") : rebalanceWeakLineEndings(wrapped, maxCharsPerLine);
    }

    private static List<String> wrapLongToken(String token, int maxCharsPerLine) {
        if (token.length() <= maxCharsPerLine) {
            return Collections.singletonList(token);
        }
        if (maxCharsPerLine <= 1) {
            List<String> chars = new ArrayList<String>();
            for (int i = 0; i < token.length(); i++) {
                chars.add(String.valueOf(token.charAt(i)));
            }
            return chars;
        }
        List<String> chunks = new ArrayList<String>();
        int index = 0;
        while (index < token.length()) {
            int end = Math.min(token.length(), index + maxCharsPerLine);
            chunks.add(token.substring(index, end));
            index = end;
        }
        return chunks;
    }

    private static List<String> rebalanceWeakLineEndings(List<String> lines, int maxCharsPerLine) {
        if (lines.size() <= 1) {
            return lines;
        }
        List<String> rebalanced = new ArrayList<String>(lines);
        int index = 0;
        while (index < rebalanced.size() - 1) {
            List<String> words = nonBlankWords(rebalanced.get(index));
            int carryCount = weakTrailingWordCount(words);
            if (carryCount > 0 && carryCount < words.size()) {
                rebalanced.set(index, join(words.subList(0, words.size() - carryCount)));
                String carried = join(words.subList(words.size() - carryCount, words.size()));
                rebalanced.set(index + 1, (carried + " " + rebalanced.get(index + 1)).trim());
            }
            carryOverflowingWordsForward(rebalanced, index + 1, maxCharsPerLine);
            index++;
        }
        return rebalanced;
    }

    private static void carryOverflowingWordsForward(List<String> lines, int startIndex, int maxCharsPerLine) {
        int index = startIndex;
        while (index < lines.size()) {
            List<String> words = nonBlankWords(lines.get(index));
            if (lines.get(index).length() <= maxCharsPerLine || words.size() <= 1) {
                index++;
                continue;
            }
            String carried = words.get(words.size() - 1);
            lines.set(index, join(words.subList(0, words.size() - 1)));
            if (index == lines.size() - 1) {
                lines.add(carried);
            } else {
                lines.set(index + 1, (carried + " " + lines.get(index + 1)).trim());
            }
            if (lines.get(index).length() <= maxCharsPerLine) {
                index++;
            }
        }
    }

    private static int weakTrailingWordCount(List<String> words) {
        if (words.isEmpty()) {
            return 0;
        }
        if (words.size() >= 2 && "as".equalsIgnoreCase(words.get(words.size() - 2))) {
            String last = stripTrailingPunctuation(words.get(words.size() - 1));
            if ("the".equalsIgnoreCase(last) || "a".equalsIgnoreCase(last) || "an".equalsIgnoreCase(last)) {
                return 2;
            }
        }
        String last = stripTrailingPunctuation(words.get(words.size() - 1)).toLowerCase();
        return WEAK_LINE_ENDINGS.contains(last) ? 1 : 0;
    }

    private static String truncateLineToFit(String text, int maxVisibleChars, int longTokenBreakLength) {
        String normalized = text == null ? "" : text.trim();
        if (maxVisibleChars <= 0) {
            return "";
        }
        if (normalized.length() <= maxVisibleChars) {
            return normalized;
        }
        if (maxVisibleChars <= 3) {
            return normalized.substring(0, Math.min(maxVisibleChars, normalized.length()));
        }

        List<String> words = nonBlankWords(normalized);
        if (words.size() <= 1) {
            return normalized.substring(0, Math.min(Math.max(1, maxVisibleChars - 3), normalized.length())) + "...";
        }

        List<String> kept = new ArrayList<String>();
        for (int i = 0; i < words.size(); i++) {
            String word = words.get(i);
            List<String> candidateWords = new ArrayList<String>(kept);
            candidateWords.add(word);
            String candidate = join(candidateWords);
            if (candidate.length() + 4 <= maxVisibleChars) {
                kept.add(word);
                continue;
            }
            if (kept.isEmpty() || word.length() >= longTokenBreakLength) {
                String partial = kept.isEmpty() ? word : join(candidateWords);
                return partial.substring(0, Math.min(Math.max(1, maxVisibleChars - 3), partial.length())) + "...";
            }
            break;
        }

        if (kept.isEmpty()) {
            return normalized.substring(0, Math.min(Math.max(1, maxVisibleChars - 3), normalized.length())) + "...";
        }
        return join(kept) + " ...";
    }

    private static List<String> nonBlankWords(String text) {
        if (text == null || text.trim().isEmpty()) {
            return Collections.emptyList();
        }
        return new ArrayList<String>(Arrays.asList(text.trim().split("\\s+")));
    }

    private static String stripTrailingPunctuation(String text) {
        return text == null ? "" : text.replaceAll("[,.;:]+$", "");
    }

    private static String join(List<String> words) {
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < words.size(); i++) {
            if (i > 0) {
                result.append(' ');
            }
            result.append(words.get(i));
        }
        return result.toString();
    }

    private static String joinLines(List<String> lines) {
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < lines.size(); i++) {
            if (i > 0) {
                result.append('\n');
            }
            result.append(lines.get(i));
        }
        return result.toString();
    }
}

File: src/main/kotlin/weaponsprocurement/ui/WimGuiText.kt
package weaponsprocurement.ui

import java.util.Locale

object WimGuiText {
    private val weakLineEndings: Set<String> = setOf(
        "a", "an", "and", "as", "at", "by", "for", "from", "in", "of", "or", "the", "to", "with",
    )

    @JvmStatic
    fun fit(text: String?, maxVisibleChars: Int): String = truncateLineToFit(text ?: "", maxVisibleChars, 10)

    @JvmStatic
    fun estimatedChars(width: Float): Int = estimatedChars(width, 0f, WimGuiStyle.TEXT_APPROX_CHAR_WIDTH)

    @JvmStatic
    fun estimatedChars(width: Float, horizontalPadding: Float, approxCharWidth: Float): Int {
        val effectiveWidth = maxOf(8f, width - maxOf(0f, horizontalPadding))
        return maxOf(1, (effectiveWidth / maxOf(1f, approxCharWidth)).toInt())
    }

    @JvmStatic
    fun wrap(text: String?, maxCharsPerLine: Int, maxLines: Int): List<String> {
        val safeChars = maxOf(1, maxCharsPerLine)
        val safeLines = maxOf(1, maxLines)
        val lines = ArrayList<String>()
        for (rawLine in (text ?: "").split(Regex("\\r?\\n"), limit = 0)) {
            lines.addAll(wrapTextLineWordAware(rawLine, safeChars))
        }
        if (lines.isEmpty()) {
            lines.add("")
        }
        if (lines.size <= safeLines) {
            return lines
        }
        val visible = ArrayList(lines.subList(0, safeLines))
        val overflow = StringBuilder()
        for (i in safeLines - 1 until lines.size) {
            if (overflow.isNotEmpty()) {
                overflow.append(' ')
            }
            overflow.append(lines[i])
        }
        visible[visible.size - 1] = fit(overflow.toString(), safeChars)
        return visible
    }

    @JvmStatic
    fun fitLayout(
        text: String?,
        availableWidth: Float,
        availableHeight: Float,
        minRowHeight: Float,
        horizontalPadding: Float,
        verticalPadding: Float,
        approxCharWidth: Float,
        lineHeight: Float,
        maxLines: Int,
        canGrowHeight: Boolean,
        maxHeight: Float,
    ): WimGuiTextLayout {
        val safeMaxLines = maxOf(1, maxLines)
        val safeLineHeight = maxOf(1f, lineHeight)
        val maxCharsPerLine = estimatedChars(availableWidth, horizontalPadding, approxCharWidth)
        var heightLimitedLines = safeMaxLines
        if (!canGrowHeight && !availableHeight.isInfinite()) {
            heightLimitedLines = maxOf(1, ((availableHeight - verticalPadding) / safeLineHeight).toInt())
        }
        val cappedMaxLines = maxOf(1, minOf(safeMaxLines, heightLimitedLines))
        val wrappedLines = wrapAll(text, maxCharsPerLine)
        val truncated = wrappedLines.size > cappedMaxLines
        val cappedLines = ArrayList<String>()
        if (truncated) {
            cappedLines.addAll(wrappedLines.subList(0, cappedMaxLines))
            val overflow = StringBuilder()
            for (i in cappedMaxLines - 1 until wrappedLines.size) {
                if (overflow.isNotEmpty()) {
                    overflow.append(' ')
                }
                overflow.append(wrappedLines[i])
            }
            cappedLines[cappedLines.size - 1] = fit(overflow.toString(), maxCharsPerLine)
        } else {
            cappedLines.addAll(wrappedLines)
        }
        if (cappedLines.isEmpty()) {
            cappedLines.add("")
        }

        val lineCount = maxOf(1, cappedLines.size)
        val desiredHeight = maxOf(minRowHeight, verticalPadding + lineCount * safeLineHeight)
        val rowHeight = if (canGrowHeight) {
            minOf(desiredHeight, maxHeight)
        } else {
            minOf(desiredHeight, availableHeight)
        }
        var widest = 0
        for (line in cappedLines) {
            widest = maxOf(widest, line.length)
        }
        return WimGuiTextLayout(
            joinLines(cappedLines),
            lineCount,
            rowHeight,
            minOf(maxOf(8f, availableWidth - horizontalPadding), widest * approxCharWidth),
            rowHeight,
            truncated,
        )
    }

    @JvmStatic
    fun fitLayout(text: String?, width: Float, minRowHeight: Float, maxLines: Int): WimGuiTextLayout = fitLayout(
        text,
        width,
        Float.POSITIVE_INFINITY,
        minRowHeight,
        WimGuiStyle.TEXT_LEFT_PAD,
        WimGuiStyle.TEXT_VERTICAL_PADDING,
        WimGuiStyle.TEXT_APPROX_CHAR_WIDTH,
        WimGuiStyle.TEXT_LINE_HEIGHT,
        maxLines,
        true,
        Float.POSITIVE_INFINITY,
    )

    private fun wrapAll(text: String?, maxCharsPerLine: Int): List<String> {
        val safeChars = maxOf(1, maxCharsPerLine)
        val lines = ArrayList<String>()
        for (rawLine in (text ?: "").split(Regex("\\r?\\n"), limit = 0)) {
            lines.addAll(wrapTextLineWordAware(rawLine, safeChars))
        }
        return if (lines.isEmpty()) listOf("") else lines
    }

    private fun wrapTextLineWordAware(line: String?, maxCharsPerLine: Int): List<String> {
        if (line == null || line.trim().isEmpty()) {
            return listOf("")
        }
        if (!line.contains(" ")) {
            return wrapLongToken(line.trim(), maxCharsPerLine)
        }

        val wrapped = ArrayList<String>()
        var current = ""
        val words = line.trim().split(Regex("\\s+"))
        for (word in words) {
            val segments = if (word.length > maxCharsPerLine) {
                wrapLongToken(word, maxCharsPerLine)
            } else {
                listOf(word)
            }
            for (segmentIndex in segments.indices) {
                val segment = segments[segmentIndex]
                current = if (current.isEmpty()) {
                    segment
                } else if (current.length + 1 + segment.length <= maxCharsPerLine) {
                    "$current $segment"
                } else {
                    wrapped.add(current)
                    segment
                }
                if (segmentIndex < segments.size - 1) {
                    wrapped.add(current)
                    current = ""
                }
            }
        }
        if (current.isNotEmpty()) {
            wrapped.add(current)
        }
        return if (wrapped.isEmpty()) listOf("") else rebalanceWeakLineEndings(wrapped, maxCharsPerLine)
    }

    private fun wrapLongToken(token: String, maxCharsPerLine: Int): List<String> {
        if (token.length <= maxCharsPerLine) {
            return listOf(token)
        }
        if (maxCharsPerLine <= 1) {
            val chars = ArrayList<String>()
            for (char in token) {
                chars.add(char.toString())
            }
            return chars
        }
        val chunks = ArrayList<String>()
        var index = 0
        while (index < token.length) {
            val end = minOf(token.length, index + maxCharsPerLine)
            chunks.add(token.substring(index, end))
            index = end
        }
        return chunks
    }

    private fun rebalanceWeakLineEndings(lines: List<String>, maxCharsPerLine: Int): List<String> {
        if (lines.size <= 1) {
            return lines
        }
        val rebalanced = ArrayList(lines)
        var index = 0
        while (index < rebalanced.size - 1) {
            val words = nonBlankWords(rebalanced[index])
            val carryCount = weakTrailingWordCount(words)
            if (carryCount > 0 && carryCount < words.size) {
                rebalanced[index] = join(words.subList(0, words.size - carryCount))
                val carried = join(words.subList(words.size - carryCount, words.size))
                rebalanced[index + 1] = "$carried ${rebalanced[index + 1]}".trim()
            }
            carryOverflowingWordsForward(rebalanced, index + 1, maxCharsPerLine)
            index++
        }
        return rebalanced
    }

    private fun carryOverflowingWordsForward(lines: MutableList<String>, startIndex: Int, maxCharsPerLine: Int) {
        var index = startIndex
        while (index < lines.size) {
            val words = nonBlankWords(lines[index])
            if (lines[index].length <= maxCharsPerLine || words.size <= 1) {
                index++
                continue
            }
            val carried = words[words.size - 1]
            lines[index] = join(words.subList(0, words.size - 1))
            if (index == lines.size - 1) {
                lines.add(carried)
            } else {
                lines[index + 1] = "$carried ${lines[index + 1]}".trim()
            }
            if (lines[index].length <= maxCharsPerLine) {
                index++
            }
        }
    }

    private fun weakTrailingWordCount(words: List<String>): Int {
        if (words.isEmpty()) {
            return 0
        }
        if (words.size >= 2 && words[words.size - 2].equals("as", ignoreCase = true)) {
            val last = stripTrailingPunctuation(words[words.size - 1])
            if (last.equals("the", ignoreCase = true) || last.equals("a", ignoreCase = true) || last.equals("an", ignoreCase = true)) {
                return 2
            }
        }
        val last = stripTrailingPunctuation(words[words.size - 1]).lowercase(Locale.ROOT)
        return if (weakLineEndings.contains(last)) 1 else 0
    }

    private fun truncateLineToFit(text: String?, maxVisibleChars: Int, longTokenBreakLength: Int): String {
        val normalized = text?.trim() ?: ""
        if (maxVisibleChars <= 0) {
            return ""
        }
        if (normalized.length <= maxVisibleChars) {
            return normalized
        }
        if (maxVisibleChars <= 3) {
            return normalized.substring(0, minOf(maxVisibleChars, normalized.length))
        }

        val words = nonBlankWords(normalized)
        if (words.size <= 1) {
            return normalized.substring(0, minOf(maxOf(1, maxVisibleChars - 3), normalized.length)) + "..."
        }

        val kept = ArrayList<String>()
        for (word in words) {
            val candidateWords = ArrayList(kept)
            candidateWords.add(word)
            val candidate = join(candidateWords)
            if (candidate.length + 4 <= maxVisibleChars) {
                kept.add(word)
                continue
            }
            if (kept.isEmpty() || word.length >= longTokenBreakLength) {
                val partial = if (kept.isEmpty()) word else join(candidateWords)
                return partial.substring(0, minOf(maxOf(1, maxVisibleChars - 3), partial.length)) + "..."
            }
            break
        }

        if (kept.isEmpty()) {
            return normalized.substring(0, minOf(maxOf(1, maxVisibleChars - 3), normalized.length)) + "..."
        }
        return join(kept) + " ..."
    }

    private fun nonBlankWords(text: String?): List<String> {
        if (text == null || text.trim().isEmpty()) {
            return emptyList()
        }
        return ArrayList(text.trim().split(Regex("\\s+")))
    }

    private fun stripTrailingPunctuation(text: String?): String = text?.replace(Regex("[,.;:]+$"), "") ?: ""

    private fun join(words: List<String>): String = words.joinToString(" ")

    private fun joinLines(lines: List<String>): String = lines.joinToString("\n")
}
